hare/broker.go

```diff
@@ -79,51 +71,16 @@ func (b *Broker) Start(ctx context.Context) {
 	b.once.Do(func() {
 		b.mu.Lock()
 		defer b.mu.Unlock()
-		if b.cancel != nil {
-			b.cancel()
-		}
 		b.ctx, b.cancel = context.WithCancel(ctx)
 	})
 }
 
-var (
-	errUnregistered      = errors.New("layer is unregistered")
-	errNotSynced         = errors.New("layer is not synced")
-	errFutureMsg         = errors.New("future message")
-	errRegistration      = errors.New("failed during registration")
-	errInstanceNotSynced = errors.New("instance not synchronized")
-	errClosed            = errors.New("closed")
-)
-
-func (b *Broker) validateTiming(ctx context.Context, m *Message) error {
-	msgInstID := m.Layer
-	if b.getInbox(msgInstID) != nil {
-		b.Log.WithContext(ctx).With().Debug("instance exists", msgInstID)
-		return nil
-	}
-
-	latestLayer := b.getLatestLayer()
-
-	if msgInstID.Before(latestLayer) {
-		return fmt.Errorf("%w: msg %v, latest %v", errUnregistered, msgInstID.Uint32(), latestLayer)
-	}
-
-	// current layer
-	if msgInstID == latestLayer {
-		return fmt.Errorf("%w: latest %v", errRegistration, latestLayer)
-	}
-
-	// early msg
-	if msgInstID == latestLayer.Add(1) {
-		return fmt.Errorf("%w: latest %v", errEarlyMsg, latestLayer)
-	}
-
-	// future msg
-	return fmt.Errorf("%w: msg %v, latest %v", errFutureMsg, msgInstID, latestLayer)
-}
+var errClosed = errors.New("closed")
 
 // HandleMessage separate listener routine that receives gossip messages and adds them to the priority queue.
 func (b *Broker) HandleMessage(ctx context.Context, _ p2p.Peer, msg []byte) pubsub.ValidationResult {
+	b.mu.Lock()
+	defer b.mu.Unlock()
```

Comment:
i am not comfortable with locking for the whole duration of processing the gossip msg.

for example, if we are under spam attack, among other things, `case out <- em:` can block.

Reply:
By `case out <- em:` are you referring to the point where the message is forwarded to the consensus process?

https://github.com/spacemeshos/go-spacemesh/blob/6271e39ddafa202c8061fe5143911eb0a0dff419/hare/broker.go#L225-L233

Could you go into more detail about how this spam attack would work and what the outcome would be?

Reply:
say current layer is 10. adversaries start spamming nodes with hare messages of layer 11 and the channel gets filled up. now nodes are not able to process anything due to the lock.

the old code seems to very careful with locking, as hare is performance critical in the code. i don't have a good reason to undo that.

Reply:
Ok got you, I think that we could solve that by forwarding the message in a new goroutine, E.G.:

```go
go func() {
    select { 
     case out <- iMsg: 
     case <-ctx.Done(): 
     	    return ctx.Err() 
     case <-b.ctx.Done(): 
 	    return errClosed 
     }
}()
```

What do you think?

Reply:
Actually maybe better we could use an anonymous function to wrap up to just before the call to the `pubKeyExtractor`, and then defer the unlock in the anonymous function.

Reply:
Actually maybe that wouldn't work, because we need to lock in order to safely reference `b.ctx` in `HandleMessage`

Reply:
i am sympathetic. i also prefer simpler locking scheme. i am a forever opponent for premature optimization.

maybe we should give it a try? i'd go for using a separate goroutine. 
regarding b.ctx, you just need to use errgroup (see hare.go) to manage goroutines.
and when shutting down broker, make sure all outstanding goroutines are terminated.

Reply:
I've updated it in [20fb45c](https://github.com/spacemeshos/go-spacemesh/pull/4130/commits/20fb45c3d7f4cf55bf3695b41059ee9844d25003) I think it's still safe, but would appreciate your scrutiny.